samza-core/src/main/java/org/apache/samza/util/StateTransitionUtil.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.util;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A utility class to perform complex state transitions.
+ */
+public class StateTransitionUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(StateTransitionUtil.class);
+
+  /**
+   * Atomically sets the value to the given updated value if current value is one of the expected values.
+   *
+   * @param reference the atomic reference
+   * @param expectedValues set of expected values
+   * @param update the new value
+   * @param <T> type of the atomic reference
+   *
+   * @return true if current state is one of the expected value and transition was successful; false otherwise
+   */
+  public static <T> boolean compareAndSet(AtomicReference<T> reference, Collection<T> expectedValues, T update) {
+    while (true) {
+      T currentValue = reference.get();
+      if (expectedValues.contains(currentValue)) {
+        if (reference.compareAndSet(currentValue, update)) {
+          return true;
+        }
+      } else {
+        return false;
+      }
+    }
+  }
+
+  /**
+   * Atomically sets the value to the given updated value if the black listed values does not contain current value.
+   *
+   * @param reference the atomic reference
+   * @param blacklistedValues set of blacklisted values
+   * @param update the new value
+   * @param <T> type of the atomic reference
+   *
+   * @return true if current state is not in the blacklisted values and transition was successful; false otherwise
+   */
+  public static <T> boolean compareNotInAndSet(AtomicReference<T> reference, Collection<T> blacklistedValues, T update) {
```

Comment:
Though it looks like we use atomic reference in StreamProcessor, we lock implicitly in StateTransitionUtil. From what I can understand, we've replaced synchronized lock with a busy spinning lock in this patch for performing state transition on a object.

My concern with the spinlock is that, while waiting to acquire a lock, it wastes the CPU cycles. Can you share the value-add and benefits that we get by replacing synchronized lock with the spinning lock. If we need to lock, why can't we use a explicit lock.